### PR TITLE
[TPU] Add TPU specific var VLLM_TPU_MOST_MODEL_LEN

### PR DESCRIPTION
**Pull request by Chenyaaang**
## Purpose

Add TPU specific environment variable: VLLM_TPU_MOST_MODEL_LEN. This is used to pass in the request length for most of requests, comparing with the existing variable`max_model_len` for the length of longest requests. This will benefit when `most_model_len` is much longer than `most_model_len`, and a large portion of requests fall inside `most_model_len`, such as 1% requests are 32k for `most_model_len` and 99% requests are 2k for `most_model_len`, in the following 2 ways:
1. When `max_model_len` is very large, attention kernel usually doesn't reach its best performance (because it falls outside the [tuned shape](https://github.com/pytorch/xla/blob/master/torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py#L46)), by using `most-model-len`, we can easily see perf gain
2. When `max_model_len` and `max_num_reqs` are both large, it's very likely to cause SMEM OOM, and we have to [increase the page size](https://github.com/vllm-project/vllm/blob/main/vllm/v1/attention/backends/pallas.py#L72) above the pre-set limit 256 (which also harms kernel performance). By introducing `most-model-len`, we keep the `max-num-seqs` when the requests' lengths are within `most-model-len`, and decrease `max-num-seqs` for the extra long outlier requests. 

The change is kept within tpu_model_runner, with no change in scheduler, so it doesn't interfere other systems. 

In tpu_model_runner.py, we initialize `num_reqs_most_model_len` and `num_reqs_max_model_len` for `most_model_len` and `max_model_len` condition respectively based on the max number without SMEM OOM. The scheduler's output is then routed to either `max_model_len` or `most_model_len` depending on the request lengths. The requests might be executed in more than one batch, and the final result will be concated together. To avoid recompilation, we pre-compiled the backbone with both `max_model_len` and `most_model_len`. 

## Test

The change is tested and compared to baseline, at commit dac8cc49. We started the server of llama3-8B, with `max-model-len = 32k`, `max-num-reqs = 128`, and send benchmark requests with `input-len = 1800`, `output-len=128`. We compare the compilation time and throughput with and without using `most-model-len = 2048`. 
<meta charset="utf-8"><div dir="ltr" style="margin-left:0pt;" align="left">
  | Backbone compilation time (sec) | Total init time (sec) | throughput
-- | -- | -- | --
w/o most-model-len | 76.06 | 177.76 | 5.75
w/ most-model-len | 120.36 | 222.49 | 7.46
</div>

The result shows a ~30% perf gain and the increase of compilation time is acceptable.

### CODE REVIEW

File: vllm/platforms/tpu.py [hunk 1]
```diff
@@ -122,16 +122,6 @@ def check_and_update_config(cls, vllm_config: VllmConfig) -> None:
                 PallasAttentionBackend)
             cache_config.block_size = PallasAttentionBackend.get_page_size(
                 vllm_config)  # type: ignore[assignment]
-            min_page_size = PallasAttentionBackend.get_min_page_size(
```
Comment (yaochengji) [inline vllm/platforms/tpu.py:125]: Why are these lines removed? IIUC, when `MOST_MODEL_LEN` is not set or too large, we still need this logic?

File: vllm/v1/worker/tpu_model_runner.py [hunk 2]
```diff
@@ -844,52 +897,74 @@ def execute_model(
...
+                logits = self.structured_decode(require_struct_decoding,
+                                                grammar_bitmask_padded, logits,
+                                                arange)
+            selected_token_ids = self.sample_from_logits_func(
+                logits, tpu_sampling_metadata)
+            # NOTE (NickLucche) Use the original logits (before any penalties or
+            # temperature scaling) for the top-k logprobs. We can't enforce it
+            # due to recompilations outside torch.compiled code, so just make
+            # sure `sample_from_logits` does not modify the logits in-place.
+            logprobs = self.gather_logprobs(logits, selected_token_ids) \
```
Comment (yaochengji) [inline vllm/v1/worker/tpu_model_runner.py:939]: I think we'd better use `logprobs.tolist()` here to make sure the following logic of `logprobs` happens on cpu.

**Review by yaochengji — COMMENTED** [review]
Thanks for the contribution! Left a few comments.

File: vllm/platforms/tpu.py (hunk 1 above)
Comment (Chenyaaang) [reply vllm/platforms/tpu.py:125]: I thought we want to adjust num_reqs no matter we have most_model_len or not? So I choose to decrease num-reqs in tpu_model_runner's init function.

File: vllm/v1/worker/tpu_model_runner.py (hunk 2 above)
Comment (Chenyaaang) [reply vllm/v1/worker/tpu_model_runner.py:939]: Thanks, fixed.

File: vllm/platforms/tpu.py (hunk 1 above)
Comment (yaochengji) [reply vllm/platforms/tpu.py:125]: Thanks, makes sense to me!

File: vllm/v1/attention/backends/pallas.py [hunk 3]
```diff
@@ -77,6 +77,11 @@ def get_min_page_size(vllm_config: VllmConfig) -> int:
         min_page_size = 1 << (min_page_size - 1).bit_length()
         return min_page_size
 
+    @staticmethod
+    def get_max_num_seqs(model_len: int, page_size: int) -> int:
+        num_page_per_req = cdiv(model_len, page_size)
+        return 1024 * 1024 // 2 // num_page_per_req // 4
```
Comment (vanbasten23) [inline vllm/v1/attention/backends/pallas.py:83]: could you explain to me this `1024 * 1024 // 2 // num_page_per_req // 4` line? What are the 1024, 2, and 4 here?

Comment (Chenyaaang) [reply vllm/v1/attention/backends/pallas.py:83]: It's same as the function`get_min_page_size` above. SMEM is 1M so 1024 * 1024, we want to allocate half of SMEM so //2, and each page index is int32 so 4 bytes.

**Review by yaochengji — APPROVED** [review]
LGTM, thanks!

File: vllm/v1/attention/backends/pallas.py (hunk 3 above)
Comment (vanbasten23) [reply vllm/v1/attention/backends/pallas.py:83]: i figured it out. 1024*1024 is the smem size for v6e. 4 is the byte of int32. I forgot what the 2 is for tho
